presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PredicatePushDown.java

```diff
@@ -211,17 +216,76 @@ public PlanNode visitProject(ProjectNode node, RewriteContext<Expression> contex
+            // Push down conjuncts from the inherited predicate that only depend on deterministic assignments with
+            // certain limitations.
+            List<Expression> deterministicConjuncts = conjuncts.get(true);
 
-            // All non-deterministic conjuncts, if any, will be in the filter node.
+            // We partition the expressions in the deterministicConjuncts into two lists, and only inline the
+            // expressions that are in the inlining targets list.
+            Map<Boolean, List<Expression>> partitionedConjuncts = partitionConjuncts(deterministicConjuncts, node);
+            List<Expression> inlinedDeterministicConjuncts = partitionedConjuncts.get(false).stream()
+                    .map(entry -> inlineSymbols(node.getAssignments().getMap(), entry))
+                    .collect(Collectors.toList());
+
+            PlanNode rewrittenNode = context.defaultRewrite(node, combineConjuncts(inlinedDeterministicConjuncts));
+
+            // All deterministic conjuncts that contains non-inlining targets, and non-deterministic conjuncts,
+            // if any, will be in the filter node.
+            List<Expression> nonInliningConjuncts = partitionedConjuncts.get(true);
             if (!conjuncts.get(false).isEmpty()) {
-                rewrittenNode = new FilterNode(idAllocator.getNextId(), rewrittenNode, combineConjuncts(conjuncts.get(false)));
+                nonInliningConjuncts.addAll(conjuncts.get(false));
+            }
+
+            if (!nonInliningConjuncts.isEmpty()) {
+                rewrittenNode = new FilterNode(idAllocator.getNextId(), rewrittenNode, combineConjuncts(nonInliningConjuncts));
             }
 
             return rewrittenNode;
         }
 
+        private boolean isInliningExpression(Expression expression, ProjectNode node)
+        {
+            Set<Symbol> childOutputSet = ImmutableSet.copyOf(node.getOutputSymbols());
+
+            // candidate symbols for inlining are
+            //   1. references to simple constants
+            //   2. references to complex expressions that
+            //      a. are not inputs to try() expressions
+            //      b. appear only once across all expressions
+            // which come from the node, as opposed to an enclosing scope.
+
+            Map<Symbol, Long> dependencies = SymbolsExtractor.extractAll(expression).stream()
+                    .filter(childOutputSet::contains)
+                    .collect(Collectors.groupingBy(Function.identity(), Collectors.counting()));
+
+            // exclude any complex inputs to TRY expressions. Inlining them would potentially
+            // change the semantics of those expressions
+            Set<Symbol> tryArguments = extractTryArguments(expression).stream().collect(toSet());
+
+            long nonInliningSymbolsCount = dependencies.entrySet().stream()
```

Comment:
No need to count. You can just do:

```java
return dependencies.entrySet().stream()
    .noneMatch(...);
```

or, even better, invert the condition to make it easier to reason about and use `allMatch`:

```java
return dependencies.entrySet().stream()
     .allMatch(entry -> (entry.getValue() == 1 || node.getAssignments().get(entry.getKey()) instanceof Literal)
                            && !tryArguments.contains(entry.getKey()))
```

Reply:
oh thanks!